core/src/main/c/share/dedup.cpp

```diff
@@ -0,0 +1,644 @@
+           index_pos <= index_hi_inc) {
+        if (src[src_pos] < index[index_pos].ts) {
+            dest[0].ts = src[src_pos];
+            dest[0].i = src_pos | (1ull << 63);
+            dest++;
+            src_pos++;
+        } else if (src[src_pos] >
+                   index[index_pos].ts) {
+            dest[0] = index[index_pos];
+            dest++;
+            index_pos++;
+        } else {
+            // index_ts == src_ts
+            const uint64_t conflict_ts = src[src_pos];
+            while (index_pos <= index_hi_inc &&
+                   index[index_pos].ts == conflict_ts) {
+                index_pos++;
+            }
+
+            // replace all records with same timestamp with last version from index
+            while (src_pos <= src_hi_incl && src[src_pos] == conflict_ts) {
+                dest[0] = index[index_pos - 1];
+                dest++;
+                src_pos++;
+            }
+        }
+    }
+
+    while (index_pos <= index_hi_inc) {
+        dest[0] = index[index_pos];
+        dest++;
+        index_pos++;
+    }
+
+    while (src_pos <= src_hi_incl) {
+        dest[0].ts = src[src_pos];
+        dest[0].i = src_pos | (1ull << 63);
+        dest++;
+        src_pos++;
+    }
+    return dest - reinterpret_cast<index_t *> (pDestIndex);
+}
+
+JNIEXPORT jlong JNICALL
+Java_io_questdb_std_Vect_mergeDedupTimestampWithLongIndexIntKeys(
+        JAVA_STATIC,
+        jlong srcTimestampAddr,
+        jlong mergeDataLo,
+        jlong mergeDataHi,
+        jlong sortedTimestampsAddr,
+        jlong mergeOOOLo,
+        jlong mergeOOOHi,
+        jlong tempIndexAddr,
+        jint dedupKeyCount,
+        jlong dedupColBuffs
+) {
+    auto *src = reinterpret_cast<uint64_t *> (srcTimestampAddr);
+    auto data_lo = __JLONG_REINTERPRET_CAST__(int64_t, mergeDataLo);
+    auto data_hi = __JLONG_REINTERPRET_CAST__(int64_t, mergeDataHi);
+    auto *index = reinterpret_cast<index_t *> (sortedTimestampsAddr);
+    auto index_lo = __JLONG_REINTERPRET_CAST__(int64_t, mergeOOOLo);
+    auto index_hi = __JLONG_REINTERPRET_CAST__(int64_t, mergeOOOHi);
+    auto *index_tmp = reinterpret_cast<index_t *> (tempIndexAddr);
+
+    const auto compare_by_rowid = [&]<class T>(const dedup_column_t<T> *column, const int64_t &col_index,
+                                               const int64_t &index_index) -> int {
+        // negative index means value is from colum, positive from get_o3_data
+        const auto l_val =
+                col_index >= column->column_top ? column->get_column_data()[col_index] : column->get_null_value();
+        const auto r_val = column->get_o3_data()[index_index];
+
+        // One of the values can be MIN of the type (null value)
+        // and subtraction can result in type overflow
+        return l_val > r_val ? 1 : (l_val < r_val ? -1 : 0);
+    };
+
+    const auto src_keys = reinterpret_cast<const dedup_column *>(dedupColBuffs);
+    if (dedupKeyCount == 1) {
```

Comment:
I'm unsure about this "unrolling" currenlty. Maybe it's better to put it back into loop.

Reply:
It is a faster comparison method with no loops, and no switches. And it's already done, I'm reluctant to remove it now.